### PR TITLE
board/feather-stm32f405: add new target board feather-stm32f405

### PR DESCRIPTION
**Pull request by ardnew**
##### These changes add support for the Adafruit STM32F405 Feather Express - no peripheral support - this is only the basic board support package

As requested to make smaller PRs, this is a subset of changes originally proposed in PR #1322.

Specifically, these changes include a new `stm32f405` machine and runtime (based on `stm32f4` build tag in PR #1332). Additionally, to actually make use of these new sources, a board implementation (and associated target file) was added for the `feather-stm32f405`.

## Notes

The exact chip on this board is `STM32F405RG`. 

A substantial amount of time was spent verifying and documenting the `runtime_stm32f405.go` source to ensure the correct bootup sequence was performed and supported chip features were enabled.

#### Programming

Flashing this board using its DFU bootloader is a bit awkward/cumbersome (see Adafruit docs linked below), but it is possible — without requiring an external programmer — by pulling the `B0` pin high (you can use the board's 3.3V output pin) at bootup when connected to host PC via USB. This puts the device in bootloader mode.

Once in bootloader mode, I've verified the device can be programmed using either `STM32CubeProgrammer` (requires ***Oracle*** JDK 8) or the open-source tool `dfu-util` (available in most package manager repos, verified using Ubuntu 20.04).

> Alternatively, I've been using a J-Link Mini EDU SWD programmer connected to the header on the reverse side of the board without issue.

|Programmer|URL|
|:-------------:|:----|
|`STM32CubeProgrammer`|https://www.st.com/en/development-tools/stm32cubeprog.html|
|`dfu-util`|http://dfu-util.sourceforge.net/|

#### ST
|Title|URL|
|:----:|:----|
|Overview|https://www.st.com/content/st_com/en/products/microcontrollers-microprocessors/stm32-32-bit-arm-cortex-mcus/stm32-high-performance-mcus/stm32f4-series/stm32f405-415/stm32f405rg.html|
|Datasheet|https://www.st.com/resource/en/datasheet/stm32f405rg.pdf|
|UserManual|https://www.st.com/resource/en/reference_manual/dm00031020-stm32f405415-stm32f407417-stm32f427437-and-stm32f429439-advanced-armbased-32bit-mcus-stmicroelectronics.pdf|

#### Adafruit

|Title|URL|
|:----:|:----|
|Overview|https://learn.adafruit.com/adafruit-stm32f405-feather-express/overview|
|Bootloader|https://learn.adafruit.com/adafruit-stm32f405-feather-express/dfu-bootloader-details|
|Schematic|https://learn.adafruit.com/adafruit-stm32f405-feather-express/downloads|

#### STM32duino 

A lot of info was derived from these relevant chip/board-specific implementations in [Arduino_Core_STM32](https://github.com/stm32duino/Arduino_Core_STM32)

|Title|URL|
|:----:|:----|
|Variant|https://github.com/stm32duino/Arduino_Core_STM32/tree/master/variants/FEATHER_F405|
|HAL|https://github.com/stm32duino/Arduino_Core_STM32/tree/master/system/STM32F4xx|

### CODE REVIEW

**Comment by deadprogram** [comment]
Hello @ardnew can you please rebase this branch against `dev` so you can resolve the merge conflict? Thank you.

**Comment by ardnew** [comment]
Unfortunately, there is code in this PR declaring UART and SPI interfaces (in `board_feather-stm32f405.go`) and configuring them (in `machine_stm32f405.go`), as these source files were originally duplicated from the existing `board_stm32f4disco.go` and `machine_stm32f405.go`. However, the runtime code in `runtime_stm32f405.go` does not actually utilize these interfaces in any way, so they aren't being used. 

The alternative would be to remove this code, but it would impact other sources (used by unrelated targets) to exclude this new board/target in their build constraints. I chose to leave the code, which seemed less risky in my opinion, especially since I have working branches for  both UART and SPI that simply haven't been submitted for PR yet until this gets merged.

Please let me know if you would rather I pull this code out of this branch, and then reintroduce it with the respective PRs implementing these two peripherals.

**Comment by ardnew** [comment]
removed UART/SPI/I2C peripheral code that was copied over from initial stm32f407 implementation. skeleton type definitions and several private functions required by `machine` package have been stubbed out so that the target will build. these will be elaborated in future PRs.

File: targets/feather-stm32f405.json
Comment (sago35) [inline targets/feather-stm32f405.json]: Please add the following

```
  "openocd-interface": "stlink-v2",
  "openocd-target": "stm32f4x"
```

If you have the following debugger, you will be able to flash with the following command.

```
$ tinygo flash --target feather-stm32f405 --size short --programmer jlink ./src/examples/blinky1
$ tinygo flash --target feather-stm32f405 --size short --programmer st-link-v2-1 ./src/examples/blinky1
$ tinygo flash --target feather-stm32f405 --size short --programmer cmsis-dap ./src/examples/blinky1
```

Of course, if you don't specify --programmer, you can write from dfu-util.

File: targets/stm32f405.ld
Comment (sago35) [inline targets/stm32f405.ld:9]: It's exactly the same as stm32f407.ld, so I think it's better to unify it.
Any reason why it would be better to separate them?

**Comment by sago35** [comment]
> src/runtime/runtime_stm32f405.go

We have not been able to confirm that the register settings are correct.
But from what I've written to feather-stm32f405, it seems to be working fine.

File: targets/feather-stm32f405.json
Comment (ardnew) [reply targets/feather-stm32f405.json]: sweet, I didn't know I could flash and debug via jlink straight from tinygo. I've been building + flashing + debugging as separate manual steps.

You do realize the feather does not have an STLink onboard, right? So would I still use that same `stlink-v2` value for the `openocd-interface` key in the json? 

And btw, for my new-ish version of OpenOCD, it ships with a bare `stlink` interface script that can detect the target interface version automatically, avoiding the need to specify, e.g., `-v2`

File: targets/stm32f405.ld
Comment (ardnew) [reply targets/stm32f405.ld:9]: The original script I used had quite a bit more configuration, but I replaced it with the `stm32f407.ld` for simplicity - to define only the bare minimum. I would personally prefer keeping it there for now and "upgrading" it after (if) the board gets merged with the original changes I wanted to include the first time.

File: targets/feather-stm32f405.json
Comment (sago35) [reply targets/feather-stm32f405.json]: Yes, I know that stlink is not on feather-stm32f405.
And I know that since 2016 or so, I've been using interface/stlink.cfg.

Maybe you can set the value to anything, so let's set it to jlink.

```
"openocd-interface": "jlink",
```

File: targets/stm32f405.ld
Comment (sago35) [reply targets/stm32f405.ld:9]: ok.

File: targets/feather-stm32f405.json
Comment (ardnew) [reply targets/feather-stm32f405.json]: I'm getting errors when trying to flash using this method, and I'm not very familiar with OpenOCD, maybe you can spot what I'm doing wrong? I'll continue researching the issue in the mean time.

```
andrew@trip /usr/local/src/tinygo/dev (feature/feather-stm32f405-i2c *)
⧉  cat targets/feather-stm32f405.json
{
  "inherits": ["cortex-m4"],
  "build-tags": ["feather_stm32f405", "stm32f405", "stm32f4", "stm32"],
  "cflags": [
    "-Qunused-arguments"
  ],
  "linkerscript": "targets/stm32f405.ld",
  "extra-files": [
    "src/device/stm32/stm32f405.s"
  ],
  "openocd-interface": "jlink",
  "openocd-target": "stm32f4x"
}
```
 
And when flashing the board:
```
andrew@trip /usr/local/src/tinygo/dev (feature/feather-stm32f405-i2c *)
⧉  tinygo flash --target feather-stm32f405 --size short --programmer jlink ./src/examples/blinky1
   code    data     bss |   flash     ram
   2952       4    4144 |    2956    4148
Open On-Chip Debugger 0.10.0+dev-01365-g583a65644-dirty (2020-07-31-09:26)
Licensed under GNU GPL v2
For bug reports, read
http://openocd.org/doc/doxygen/bugs.html
Info : auto-selecting first available session transport "jtag". To override use 'transport select <transport>'.
Info : J-Link EDU Mini V1 compiled Jul 17 2020 16:25:21
Info : Hardware version: 1.00
Info : VTarget = 3.292 V
Info : clock speed 2000 kHz
Error: JTAG scan chain interrogation failed: all ones
Error: Check JTAG interface, timings, target power, etc.
Error: Trying to use configured scan chain anyway...
Error: stm32f4x.cpu: IR capture error; saw 0x0f not 0x01
Warn : Bypassing JTAG setup events due to errors
Error: Invalid ACK (7) in DAP response
Error: JTAG-DP STICKY ERROR
Error: Invalid ACK (7) in DAP response
Error: JTAG-DP STICKY ERROR
  -- *snip* repeats many times --
Error: Invalid ACK (7) in DAP response
Error: JTAG-DP STICKY ERROR
Error: Invalid ACK (7) in DAP response
Error: JTAG-DP STICKY ERROR
in procedure 'program'
** OpenOCD init failed **
shutdown command invoked

error: failed to flash /tmp/tinygo111238838/main.hex: exit status 1
```

Comment (aykevl) [reply targets/feather-stm32f405.json]:
```
Info : auto-selecting first available session transport "jtag". To override use 'transport select <transport>'.
```

I think you need an extra `"openocd-transport": "swd"` in the target JSON file, to set the transport to SWD (instead of JTAG). Most Cortex-M chips use SWD.

Comment (ardnew) [reply targets/feather-stm32f405.json]: awesome -- adding `"openocd-transport": "swd"` worked great!

and right you are, had to solder to the SWD pins on this board, it don't believe it even has JTAG

going to update this with your all's suggestions

```
andrew@trip ~/Code/tinygo/dev (feature/feather-stm32f405-i2c *)
⧉  tinygo flash --target feather-stm32f405 --size short --programmer jlink ./src/examples/blinky1
   code    data     bss |   flash     ram
   2908       0    4144 |    2908    4144
Open On-Chip Debugger 0.10.0+dev-01365-g583a65644-dirty (2020-07-31-09:26)
Licensed under GNU GPL v2
For bug reports, read
http://openocd.org/doc/doxygen/bugs.html
swd
Info : J-Link EDU Mini V1 compiled Jul 17 2020 16:25:21
Info : Hardware version: 1.00
Info : VTarget = 3.288 V
Info : clock speed 2000 kHz
Info : SWD DPIDR 0x2ba01477
Info : stm32f4x.cpu: hardware has 6 breakpoints, 4 watchpoints
Info : starting gdb server for stm32f4x.cpu on 3333
Info : Listening on port 3333 for gdb connections
target halted due to debug-request, current mode: Thread
xPSR: 0x01000000 pc: 0x08000900 msp: 0x20001000
Info : Reduced speed from 8000 kHz to 4000 kHz (maximum).
** Programming Started **
Info : device id = 0x10076413
Info : flash size = 1024 kbytes
** Programming Finished **
** Resetting Target **
shutdown command invoked
```

File: targets/feather-stm32f405.json
Comment (sago35) [inline targets/feather-stm32f405.json:17]: I think the default writing method should be dfu-util.

I added flash-command and confirmed that it can flash by dfu-util.
The reason you can''t flash properly originally is because it was {hex}.

```
{
  "inherits": ["cortex-m4"],
  "build-tags": ["feather_stm32f405", "stm32f405", "stm32f4", "stm32"],
  "automatic-stack-size": false,
  "default-stack-size": 1024,
  "cflags": [
    "-Qunused-arguments"
  ],
  "linkerscript": "targets/stm32f405.ld",
  "extra-files": [
    "src/device/stm32/stm32f405.s"
  ],
  "flash-method": "command",
  "flash-command": "dfu-util -a 0 --dfuse-address 0x08000000 -D {bin}",
  "openocd-transport": "swd",
  "openocd-interface": "jlink",
  "openocd-target": "stm32f4x"
}
```

You can flash in any of the following ways

```
$ tinygo flash -target feather-stm32f405 -size short ./src/examples/blinky1
$ tinygo flash -target feather-stm32f405 -size short -programmer jlink ./src/examples/blinky1
$ tinygo flash -target feather-stm32f405 -size short -programmer cmsis-dap ./src/examples/blinky1
```

Comment (sago35) [reply targets/feather-stm32f405.json:17]: The following is better.

```
  "flash-command": "dfu-util --alt 0 --dfuse-address 0x08000000 --download {bin}",
```

**Comment by deadprogram** [comment]
Are there further changes being requested here, or is this PR ready?

**Comment by ardnew** [comment]
@deadprogram just pushed (dbcff34) the last change requested by @sago35. 

i'm not aware of any other changes requested, so should be ready to merge once CI checks complete

File: targets/feather-stm32f405.json
Comment (sago35) [reply targets/feather-stm32f405.json:17]: The latest commit confirms it.
Very good.

```
$ tinygo flash --target feather-stm32f405 --size short ./src/examples/b1
   code    data     bss |   flash     ram
   2908       0    4144 |    2908    4144
dfu-util 0.9

Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
Copyright 2010-2016 Tormod Volden and Stefan Schmidt
This program is Free Software and has ABSOLUTELY NO WARRANTY
Please report bugs to http://sourceforge.net/p/dfu-util/tickets/

dfu-util: Invalid DFU suffix signature
dfu-util: A valid DFU suffix will be required in a future dfu-util release!!!
Opening DFU capable USB device...
ID 0483:df11
Run-time device DFU version 011a
Claiming USB DFU Interface...
Setting Alternate Setting #0 ...
Determining device status: state = dfuERROR, status = 10
dfuERROR, clearing status
Determining device status: state = dfuIDLE, status = 0
dfuIDLE, continuing
DFU mode device DFU version 011a
Device returned transfer size 2048
DfuSe interface name: "Internal Flash  "
Downloading to address = 0x08000000, size = 2908
Download	[=========================] 100%         2908 bytes
Download done.
File downloaded successfully
```

```
$ tinygo flash --target feather-stm32f405 --size short --programmer cmsis-dap ./src/examples/b5
   code    data     bss |   flash     ram
   2908       0    4144 |    2908    4144
Open On-Chip Debugger 0.10.0+dev-g5c9350d2 (2020-08-07-12:56)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
swd
Info : Using CMSIS-DAPv2 interface with VID:PID=0x6666:0x6666, serial=123456
Info : CMSIS-DAP: SWD  Supported
Info : CMSIS-DAP: FW Version = v0.1
Info : CMSIS-DAP: Serial# = 123456
Info : CMSIS-DAP: Interface Initialised (SWD)
Info : SWCLK/TCK = 1 SWDIO/TMS = 1 TDI = 1 TDO = 1 nTRST = 1 nRESET = 1
Info : CMSIS-DAP: Interface ready
Info : clock speed 2000 kHz
Info : SWD DPIDR 0x2ba01477
Info : stm32f4x.cpu: hardware has 6 breakpoints, 4 watchpoints
Info : starting gdb server for stm32f4x.cpu on 3333
Info : Listening on port 3333 for gdb connections
target halted due to debug-request, current mode: Thread 
xPSR: 0x01000000 pc: 0x0800059a msp: 0x20001000
Info : High speed (adapter speed 8000) may be limited by adapter firmware.
** Programming Started **
Info : device id = 0x10076413
Info : flash size = 1024 kbytes
** Programming Finished **
** Verify Started **
** Verified OK **
** Resetting Target **
shutdown command invoked

```

**Review by sago35 — APPROVED** [review]
LGTM

**Comment by deadprogram** [comment]
Thank you very much @ardnew for all your fine efforts on this contribution, and also thanks @sago35 and @aykevl for helping review.

OK, I am going to merge. Thanks again!

**Pull request merged**
